### PR TITLE
Bugfix consume_prefix_in_state_dict_if_present function to keep the order of the state_dict

### LINKED ISSUE #106942: [Minor Bug] Should consume_prefix_in_state_dict_if_present change ordering of keys?
### 🐛 Describe the bug

Hi, I was looking into the source code for `torch.nn.modules.utils.consume_prefix_in_state_dict_if_present` and saw that the very first thing it did was to sort the keys, which would change the ordering of the keys downstream during inplace assignment of new prefix-removed keys.

Perhaps I overlooked something, but I was wondering if this is intended behaviour, because it is not mentioned in the docstring.
The model would load fine with `load_state_dict`, but with `nn.Module` state dicts being OrderedDicts, just seems a little uncessary and confusing.

Example code:
```
# python 3.8.10, torch 2.0.1+cu117
from collections import OrderedDict
from torch.nn.modules.utils import consume_prefix_in_state_dict_if_present

state_dict = OrderedDict()
state_dict['module.body.stage1.0.0.weight'] = 0
state_dict['module.body.stage1.0.1.weight'] = 1
state_dict['module.body.stage1.0.1.bias'] = 2
state_dict['module.bboxes.0.conv1x1.weight'] = 3
state_dict['module.bboxes.0.conv1x1.bias'] = 4

for k in state_dict:
    print(k)
print("-------------------------")

# now the ordering is changed
consume_prefix_in_state_dict_if_present(state_dict, 'module.')
for k in state_dict:
    print(k)


### Versions

Collecting environment information...
PyTorch version: 2.0.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.27.0
Libc version: glibc-2.31

Python version: 3.8.10 (default, Mar 15 2022, 12:22:08)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.4.0-117-generic-x86_64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: 11.6.124
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-80GB
GPU 1: NVIDIA A100-SXM4-80GB
GPU 2: NVIDIA A100-SXM4-80GB
GPU 3: NVIDIA A100-SXM4-80GB
GPU 4: NVIDIA A100-SXM4-80GB
GPU 5: NVIDIA A100-SXM4-80GB
GPU 6: NVIDIA A100-SXM4-80GB
GPU 7: NVIDIA A100-SXM4-80GB

Nvidia driver version: 510.47.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.7.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   48 bits physical, 48 bits virtual
CPU(s):                          128
On-line CPU(s) list:             0-127
Thread(s) per core:              2
Core(s) per socket:              32
Socket(s):                       2
NUMA node(s):                    2
Vendor ID:                       AuthenticAMD
CPU family:                      25
Model:                           1
Model name:                      AMD EPYC 7513 32-Core Processor
Stepping:                        1
Frequency boost:                 enabled
CPU MHz:                         1492.760
CPU max MHz:                     2600.0000
CPU min MHz:                     1500.0000
BogoMIPS:                        5190.45
Virtualization:                  AMD-V
L1d cache:                       2 MiB
L1i cache:                       2 MiB
L2 cache:                        32 MiB
L3 cache:                        256 MiB
NUMA node0 CPU(s):               0-31,64-95
NUMA node1 CPU(s):               32-63,96-127
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca

Versions of relevant libraries:
[pip3] numpy==1.24.4
[pip3] torch==2.0.1
[pip3] torchaudio==2.0.2
[pip3] torchvision==0.15.2
[pip3] triton==2.0.0
[conda] Could not collect

cc @svekars @carljparker @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

### PR DESCRIPTION
**Pull request by jmarintur**
This PR proposes to keep the original order as the original state_dict, as the issue creator proposed. It also removes a bug concerning how ``_metadata`` is handled (see below), as well as other small changes to properly remove the prefix when is present. 

In the original code, ``_metadata`` was handled as a ``key``.

```
    # also strip the prefix in metadata if any.
    if "_metadata" in state_dict:
```

This is not the case, ``_metadata`` is actually an ``attribute``. Hence, the previous condition is changed to:

```
    # also strip the prefix in metadata if any.
    if hasattr(state_dict, "_metadata"):
```

This PR also includes the necessary test.

Fixes #106942

### CODE REVIEW

**Comment by jmarintur** [comment]
Hi @mikaylagawarecki , I've solved the CI issues, could someone trigger the workflows again?

File: torch/nn/modules/utils.py [hunk 1]
```diff
@@ -57,22 +57,28 @@ def consume_prefix_in_state_dict_if_present(
...
-    for key in keys:
+    keys_to_move = list(state_dict.keys())
+    for key in keys_to_move:
+        # Key starts with the prefix
         if key.startswith(prefix):
-            newkey = key[len(prefix) :]
-            state_dict[newkey] = state_dict.pop(key)
-
-    # also strip the prefix in metadata if any.
-    if "_metadata" in state_dict:
```
Comment (mikaylagawarecki) [inline torch/nn/modules/utils.py:67]: Looking at the initial code, I am a little confused --

`_metadata` should be an attribute on the `state_dict()`

This can be seen here 
https://github.com/pytorch/pytorch/blob/af9b6fa04e7460b64e634b1aecd0ef32cd0d1ddc/torch/nn/modules/module.py#L1882-L1888

So if I understand correctly, this function is never correctly stripping the prefixes of `state_dict._metadata`. This looks like a bug to me

**Review by mikaylagawarecki — COMMENTED** [review]
See the comment regarding `_metadata`, I think the existing implementation might have a bug

Could you also add a test or update the corresponding testing for this function

File: torch/nn/modules/utils.py (hunk 1 above)
Comment (jmarintur) [reply torch/nn/modules/utils.py:67]: As you pointed out, _metadata is an attribute, and yes, it looks like a bug.

**Comment by jmarintur** [comment]
Thanks @mikaylagawarecki, I'll make the necessary changes and add / update the corresponding test.

**Comment by jmarintur** [comment]
Hi @jbschlosser, following @mikaylagawarecki suggestions, I've solved the bug and created a test for this function. 

Within the test, I considered three cases, empty state_dict, non-ddp state_dict (the prefix should not alter the state_dict) and a ddp example. In each of these cases I check that the order is preserved. 

I've also found that the key 'module' should get removed, otherwise you won't get the original dict in the _metadata.

File: torch/nn/modules/utils.py [hunk 2]
```diff
@@ -57,22 +57,25 @@ def consume_prefix_in_state_dict_if_present(
...
             # 'module': for the actual model.
             # 'module.xx.xx': for the rest.
-
             if len(key) == 0:
                 continue
-            newkey = key[len(prefix) :]
-            metadata[newkey] = metadata.pop(key)
+            # we also remove the key 'module'
+            if key == 'module':
+                state_dict._metadata.pop(key)
```
[the remark this answers is not in the document]
Comment (jmarintur) [reply torch/nn/modules/utils.py:78]: This is way:

ddp_dp_examples

Perhaps there's a better way of removing it to get back to the original _metadata.

File: test/test_nn.py [hunk 3]
```diff
@@ -12965,6 +12965,68 @@ def test_fuse_linear_bn_requires_grad(self):
             self.assertEqual(weight.requires_grad, w_rg)
             self.assertEqual(bias.requires_grad, b_rg)
 
+class TestConsumePrefix(TestCase):
```
[the remark this answers is not in the document]
Comment (jmarintur) [reply test/test_nn.py:12968]: I'll add this change, thanks!

File: test/test_nn.py [hunk 4]
```diff
@@ -12965,6 +12965,68 @@ def test_fuse_linear_bn_requires_grad(self):
...
+    # pr-117464 fixes issue-106942
```
[the remark this answers is not in the document]
Comment (jmarintur) [reply test/test_nn.py:12969]: Good, I saw it before in there, and I thought it was common practice. I'll remove it then. Thanks!

File: test/test_nn.py [hunk 5]
```diff
@@ -12965,6 +12965,68 @@ def test_fuse_linear_bn_requires_grad(self):
...
+        # 0. Case non-DDP model empty state_dict
+        net = nn.Module()
+        state_dict = net.state_dict()
+        nn.modules.utils.consume_prefix_in_state_dict_if_present(state_dict, 'module.')
+        # check they are the same preserving order
+        self.assertEqual(list(state_dict.keys()), list(net.state_dict().keys()))
+        self.assertEqual(list(state_dict._metadata.keys()), list(net.state_dict()._metadata.keys()))
+
+        # 1. Case non-DDP model test example state_dict
+        l = nn.Linear(5, 5)
```
[the remark this answers is not in the document]
Comment (jmarintur) [reply test/test_nn.py:12980]: Indeed, much better. Thanks!

File: test/test_nn.py [hunk 6]
```diff
@@ -12965,6 +12965,68 @@ def test_fuse_linear_bn_requires_grad(self):
...
+
+        # 2. Case DDP model test example state_dict
+        l = nn.Linear(5, 5)
+        bn = nn.BatchNorm2d(2)
+        conv1 = nn.Conv2d(3, 3, 3, bias=True)
+        net = nn.Module()
+        net.linear = l
+        net.bn = bn
+        net.conv1 = conv1
+        state_dict = net.state_dict()
```
[the remark this answers is not in the document]
Comment (jmarintur) [reply test/test_nn.py:13003]: I'll do that. Thanks!

File: test/test_nn.py [hunk 7]
```diff
@@ -12965,6 +12965,68 @@ def test_fuse_linear_bn_requires_grad(self):
...
+            'module.bn.num_batches_tracked': bn.num_batches_tracked,
+            'module.conv1.weight': conv1.weight,
+            'module.conv1.bias': conv1.bias,
+        })
+        ddp_state_dict._metadata.update({
+            'module': state_dict._metadata[''],
+            'module.linear': state_dict._metadata['linear'],
+            'module.bn': state_dict._metadata['bn'],
+            'module.conv1': state_dict._metadata['conv1']
+        })
```
[the remark this answers is not in the document]
Comment (jmarintur) [reply test/test_nn.py:13024]: We will need to add the key also ('metadata', {'version': 1}), don't we? To follow the same _metadata you get after DP and DDP (see my [answer](https://github.com/pytorch/pytorch/pull/117464/files/e9a73b3d3c32ce7367e8bd077945c6930030cd3a#r1475351105) below).

[the remark this answers is not in the document]
Comment (jmarintur) [reply test/test_nn.py:13024]: Since prefix is 'module.', a better option would be to do it like this:

`if key == prefix[:-1]:`

**Comment by jmarintur** [comment]
Hi @mikaylagawarecki , please check my last commit with all the changes you suggested. Please let me know if there's something else we should handle.

File: torch/nn/modules/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikaylagawarecki) [reply torch/nn/modules/utils.py:78]: Hmm I see, in the case of the DP/DDP state dict,

isn't `{'module', {'version: 1}}` the one that should map to `{'', {'version: 1}}` in the `nn.Module` that DDP wraps.

I think what we want to remove is the version of the DDP wrapper class which is the `{'', {'version: 1}}`. So this should be 

```
if key == '':
    state_dict._metadata.pop(key) 
```

then the following  `if key.startswith(prefix)` block will correctly handle transforming `{'module', {'version: 1}}` to `{'', {'version: 1}}`

**Review by mikaylagawarecki — COMMENTED** [review]
Looks good, just one more comment

**Comment by jmarintur** [comment]
Hi @mikaylagawarecki, thank you a lot for your comments. The prefix that we pass is 'module.' if we use .startswith(prefix) we won't be replacing 'module' by ''. We should do something like this, don't we? 

```
    # also strip the prefix in metadata if any.
    if hasattr(state_dict, "_metadata"):
        keys = list(state_dict._metadata.keys())
        for key in keys:
            # for the metadata dict, the key can be:
            # '': for the DDP module, which we want to remove.
            # 'module': for the actual model.
            # 'module.xx.xx': for the rest.
            if len(key) == 0:
                continue
            # handling both, 'module' case and  'module.' cases
            if key == prefix.replace('.','') or key.startswith(prefix):
                newkey = key[len(prefix) :]
                state_dict._metadata[newkey] = state_dict._metadata.pop(key)
```

**Comment by jmarintur** [comment]
I've just pushed a new commit to handle the case 'module' without dot.

**Comment by jmarintur** [comment]
Hi @mikaylagawarecki, is the last commit fine with you?

File: torch/nn/modules/utils.py [hunk 8]
```diff
@@ -57,22 +57,23 @@ def consume_prefix_in_state_dict_if_present(
...
             # 'module.xx.xx': for the rest.
-
             if len(key) == 0:
                 continue
-            newkey = key[len(prefix) :]
-            metadata[newkey] = metadata.pop(key)
+            # handling both, 'module' case and  'module.' cases
+            if key == prefix.replace('.', '') or key.startswith(prefix):
+                newkey = key[len(prefix) :]
+                state_dict._metadata[newkey] = state_dict._metadata.pop(key)
```
Comment (mikaylagawarecki) [inline torch/nn/modules/utils.py:79]: Ah, so the 

value of `state_dict._metadata['module']` will replace the `state_dict._metadata['']` is that right?

File: torch/nn/modules/utils.py [hunk 9]
```diff
@@ -57,22 +57,23 @@ def consume_prefix_in_state_dict_if_present(
...
             # '': for the DDP module, which we want to remove.
             # 'module': for the actual model.
             # 'module.xx.xx': for the rest.
-
             if len(key) == 0:
                 continue
-            newkey = key[len(prefix) :]
-            metadata[newkey] = metadata.pop(key)
+            # handling both, 'module' case and  'module.' cases
+            if key == prefix.replace('.', '') or key.startswith(prefix):
```
Comment (mikaylagawarecki) [inline torch/nn/modules/utils.py:77]: Technically could be buggy if prefix is ever something like `abc.module.`, but all uses of this function seem to pass prefix  `module.` so I think this is okay

**Review by mikaylagawarecki — APPROVED** [review]
Thanks, and apologies for the delay! This looks good to me pending green CI

The way that the function allows the prefix that is stripped be an argument instead of always using `"module."` makes this a little weird I guess, but I am happy with the change since it seems like all calls to this function seem to pass "module." as the prefix

Could you add a snippet to the PR description that describes the bugfix to `_metadata` this made please

File: torch/nn/modules/utils.py (hunk 8 above)
Comment (jmarintur) [reply torch/nn/modules/utils.py:79]: yes!

File: torch/nn/modules/utils.py (hunk 9 above)
Comment (jmarintur) [reply torch/nn/modules/utils.py:77]: Good observation, if you think it is a better solution, we could perhaps change it to ``key == prefix[:-1]``. Although, as you said, all uses of this function seems to pass prefix ``module.``.

I've already change the PR description. Thank you for your support!

**Comment by jmarintur** [comment]
Failures seem unrelated.

**Comment by jmarintur** [comment]
Hi @mikaylagawarecki, I've just seen I didn't check the "allow edits..." Sorry about that!
